### ISSUE TITLE
When running rflash on multiple node groups, with errors, the summary does not make sense.

### ISSUE THREAD

**Issue by whowutwut** [xcat2/xcat-core#5288]
Here's output from the rflash in `/var/log/xcat/commands.log`
 
```
====================================================
[Date]       2018-06-11 13:16:18
[ClientType] cli
[Request]    rflash compute,service -a 'witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar'
[Response]
Attempting to upload /firmware/Witherspoon/OP910/OP910_1742FxR9-910.1806.20180611a-pnorOnlyUpdate/witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar, please wait...
Attempting to upload /firmware/Witherspoon/OP910/OP910_1742FxR9-910.1806.20180611a-pnorOnlyUpdate/witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar, please wait...
Attempting to upload /firmware/Witherspoon/OP910/OP910_1742FxR9-910.1806.20180611a-pnorOnlyUpdate/witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar, please wait...
Attempting to upload /firmware/Witherspoon/OP910/OP910_1742FxR9-910.1806.20180611a-pnorOnlyUpdate/witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar, please wait...
sn02: Error: [500] Login to BMC failed: 500 Can't connect to 192.168.85.3:443.
sn02: Error: BMC did not respond. Validate BMC configuration and retry the command.
cn08: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn11: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn14: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn12: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn03: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn05: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn04: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn06: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn02: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn07: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn10: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn01: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn09: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn01: Error: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation failed.
sn01: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn12: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn01: Error: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation failed.
sn01: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
cn12: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn02: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn03: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn11: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn05: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn07: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn06: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn10: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn08: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
cn04: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
-------------------------------------------------------
Firmware update complete: Total=6 Success=5 Failed=1
cn01: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation failed.
-------------------------------------------------------
cn14: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
-------------------------------------------------------
Firmware update complete: Total=1 Success=1 Failed=0
-------------------------------------------------------
cn09: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
-------------------------------------------------------
Firmware update complete: Total=6 Success=6 Failed=0
-------------------------------------------------------
sn01: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
-------------------------------------------------------
Firmware update complete: Total=2 Success=1 Failed=1
sn02: BMC is not ready
-------------------------------------------------------
[ElapsedTime] 301.838 s
====================================================
```

Here's the node groups:
```
(0) root @ xcat1: /var/log/xcat/rflash
# nodels compute
cn01
cn02
cn03
cn04
cn05
cn06
cn07
cn08
cn09
cn10
cn11
cn12
cn14

(0) root @ xcat1: /var/log/xcat/rflash
# nodels service
sn01
sn02

```

Here's the service node relationship:
```
(1) root @ xcat1: /var/log/xcat/rflash
# lsdef compute -i servicenode
Object name: cn01
    servicenode=sn01,sn02
Object name: cn02
    servicenode=sn01,sn02
Object name: cn03
    servicenode=sn01,sn02
Object name: cn04
    servicenode=sn01,sn02
Object name: cn05
    servicenode=sn01,sn02
Object name: cn06
    servicenode=sn01,sn02
Object name: cn07
    servicenode=sn02,sn01
Object name: cn08
    servicenode=sn02,sn01
Object name: cn09
    servicenode=sn02,sn01
Object name: cn10
    servicenode=sn02,sn01
Object name: cn11
    servicenode=sn02,sn01
Object name: cn12
    servicenode=sn02,sn01
Object name: cn14
    servicenode=sn01
```

it looks like perhaps there two summary running for each node group... not sure why...

**Comment by gurevichmark** [comment]
I think what is happening is that with service nodes, each service node runs `rflash` on its own set of nodes. Then the output gets displayed on MN from each service node.
So these two summaries come from service nodes:
``
Firmware update complete: Total=6 Success=5 Failed=1

Firmware update complete: Total=6 Success=6 Failed=0
```

This summary is probably from rflash of SN themselves:
```
Firmware update complete: Total=2 Success=1 Failed=1
```

But I do not understand why this summary line is displayed:
```
Firmware update complete: Total=1 Success=1 Failed=0
```

**Comment by xuweibj** [comment]
In our current code, all compute nodes will be classified to 3 types: 
first type: ``sn01,sn02``
second type: ``sn02,sn01``
third type: ``sn01``

So here will receive 4 summaries. (Each type has 1, and rflash against SN has 1)
If we want to fix this, should modify the code of getting service node.

**Comment by robin2008** [comment]
No good method for this situation, as `preprocess` split those nodes into different sub-command, and no way to converge them together.

Maybe we can give more clean message like, `Firmware update complete on <xxx>: `

**Comment by gurevichmark** [comment]
@xuweibj Is there a reason why third type is not part of first type ? Just for easy parsing ?

**Comment by gurevichmark** [comment]
@robin2008 When you say: `Firmware update complete on <xxx>:` do you mean for `<xxx>` to be the name of the service node and then after `:` list all the computes that this service node processed ?

**Comment by whowutwut** [comment]
@robin2008 @xuweibj @gurevichmark Is there a way to display  the actions running on the service node, and understand where the messages are coming from, instead of looking like there's two summary on the MN node covering the same nodes...  Does the verbose output show it?

**Comment by xuweibj** [comment]
@gurevichmark , the function we called to get service node ``xCAT::ServiceNodeUtils->get_ServiceNode`` just get attribute "servicenode" from each node's definition, does not check whether has more than 1 node, consider it as a string. So "sn01,sn02", "sn02,sn01" are different. 

@whowutwut , yes run command with ``-V`` could display where the msg from. Output like this:
```
# rflash mid05tor12cn16 -a witherspoon.pnor.squashfs.tar -V
[briggs01]: Attempting to upload /mnt/xcat/iso/openbmc/910.1806.20180525a_1742Fx-r8/witherspoon.pnor.squashfs.tar, please wait...
[briggs01]: Running command in Perl
mid05tor12cn16: [briggs01]: DD 2.1 processor
mid05tor12cn16: [briggs01]: Uploading /mnt/xcat/iso/openbmc/910.1806.20180525a_1742Fx-r8/witherspoon.pnor.squashfs.tar ...
mid05tor12cn16: [briggs01]: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.167 (ID: 7c6bacc0)
mid05tor12cn16: [briggs01]: rflash IBM-witherspoon-ibm-OP9_v1.19_1.167 started, please wait...
[briggs01]: mid05tor12cn16: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.167 activation successful.
[briggs01]: -------------------------------------------------------
[briggs01]: Firmware update complete: Total=1 Success=1 Failed=0
[briggs01]: -------------------------------------------------------
```

**Comment by robin2008** [comment]
@gurevichmark @whowutwut  using `-V` could show the server name for each messages.

But for this scenario, I prefer to show the server name in the message itself without `-V`.  And now in `command.log`, we cannot see the service name. Maybe we can show them by default for better debugging.

**Comment by gurevichmark** [comment]
And maybe some more processing can be done after calling `xCAT::ServiceNodeUtils->get_ServiceNode()` so that `first type: sn01,sn02` and `third type: sn01` collapsed into a single group to reduce the summary output.

**Comment by gurevichmark** [comment]
We recently added the PR https://github.com/xcat2/xcat-core/pull/5119 to display SN for warning and error messages. Maybe this should be done for all types of messages ?

**Comment by whowutwut** [comment]
@zet809 Let's get someone to work on this usability enhancement please.

**Comment by xuweibj** [comment]
Hi @gurevichmark , if we want to collapse ``first type`` and ``third type``, we should check ``sn01`` status in ``xCAT::ServiceNodeUtils->get_ServiceNode()``. It's complex and more effort.

And more, current if we get sn01 does not work when we want to dispatch, we could dispatch command to ``sn02`` for ``first type``. But if we get sn01 worked well and collapsed them, but it's down when we want to dispatch, we will fail for all node with ``first`` type.

@whowutwut, I think we could enhance just as @gurevichmark comment, display SN for all types of message. What's your idea?

**Comment by robin2008** [comment]
It is fixed in #5401 , and for commands.log, I've logged new ticket for it.

**Comment by whowutwut** [comment]
Since this is fixed according to the comments above, closing this issue.
